Lib/pathlib.py

```diff
@@ -80,12 +77,13 @@ def _is_case_sensitive(pathmod):
 def _compile_pattern(pat, sep, case_sensitive):
     """Compile given glob pattern to a re.Pattern object (observing case
     sensitivity)."""
-    flags = re.NOFLAG if case_sensitive else re.IGNORECASE
-    regex = glob.translate(pat, recursive=True, include_hidden=True, seps=sep)
+    from glob import translate
+    regex = translate(pat, recursive=True, include_hidden=True, seps=sep)
     # The string representation of an empty path is a single dot ('.'). Empty
     # paths shouldn't match wildcards, so we consume it with an atomic group.
     regex = r'(\.\Z)?+' + regex
-    return re.compile(regex, flags).match
+    from re import compile, NOFLAG, IGNORECASE
+    return compile(regex, flags=NOFLAG if case_sensitive else IGNORECASE).match
```

Comment:
This rewrite slows down the following benchmark (which I deliberately wrote to show the danger of this kind of thing ;)

<details>
<summary>Benchmark script</summary>

```py
random_data = [
    'GgPOrkGyWTSqm', 'bfVTQFBAhIinRfaWHbq', 'DBectpWwHT', 'MMeociJaZxrCxf', 'zmggnyfHPPGljNvCQagD', 'ovlbVDMIAVQvKOS', 'KxptLDEYcyZWML', 'cFrhoqQsITgrLBmsbg', 'zjojoXVBIdbydDdN', 'qewJSKCBJWVdaVjnkQTU',
    'GhJlOoCkzdpB', 'ZtiGMhqLQojY', 'BcJHTzTJPqbBCLiI', 'DtrrRQNoXsdEQ', 'JBWFMocTwEaJtGVQWXqS', 'ytsUszUPuQrhwOeBHP', 'zKyxiAnKhLS', 'iqPbotQByCSwTxowsZ', 'qmMYujyzAXVr', 'iUWqhOLdlBaRrYs',
    'slBgBGLneo', 'SoLkUFlsEzV', 'UbePpUfojnCjIbJF', 'EgQxgPDOmwhzmLE', 'vDAWzoQwsrThN', 'TCCpLKlNXcIFDt', 'yqYXSsDqnpCl', 'RcEVhnWMlNdMbXRpw', 'PLreLUJEWiu', 'GtDrOkAqqKTh',
    'McqZmQwxuIYKgRtTOZ', 'ShEPGVcYLPVCdTjl', 'LJRkHviAVVIaJ', 'oTeiJGXVRaEXsoSHmCx', 'TCFMWcVNdEbiMHTai', 'QUQfmshWijChwT', 'kKPZAEnBizUk', 'knYGzjzVMpwuPW', 'IMUygmgTUVBNC', 'CztDXUedmKUKLFFzdyn',
    'kPCcKJGwRrEc', 'kaBxnOuxgSTUSENfLf', 'RBBzvlrrOsjFdTSrcKJy', 'caeSmoeBAClAHKVy', 'zSpoazMtjkWaTevpkIAw', 'dTuoqelDOvxfnDXl', 'WFYqVwzLAdyLkgWjMY', 'VZiNbkPKyTdSB', 'EdWGlgwlehdpwwaWQanS', 'WiwNVSwvXskswANqck',
    'fIlBHbQVdIEarTkH', 'LHAmAISOCsCOha', 'QOjCScEnLBWrzUGad', 'dNdPvTLVQZxDuXVxxOlK', 'crRorhGwfcVnSJZJRaIn', 'SLgOmawQrVujJQevrql', 'wJEuCqRSGAV', 'mVRhHMeJaHWL', 'fyEeQIFcKaJdLn', 'tnSPGVazwtzIutuHqSH',
    'ZxKweQnguxIluqRALY', 'HPGBSkhihsXtAe', 'HOKsQDZDUyF', 'WJYvIrPSKIzB', 'nOkPtVIVWTvPteIsb', 'avJtSNLNdfpYUgQJTZ', 'VfjhGvgOFjtQpLkH', 'GscNYZdgjFlKbVg', 'YdnqGKbXEJc', 'WmCjBnGrqvT',
    'LInUlnqcHqrIbx', 'dnXMyikTnuoVGN', 'nkQjWTEZGONWWReHDG', 'hxHPhRlTgP', 'EdAKHESqxQvzvlABYHf', 'gwNwwIIQLLzwvbOTqfUM', 'bMbkuVHsSN', 'wWAHjIEKsRVTDwWpRJC', 'MoWbpSGQzIYXDRSvDEG', 'UDDlCLaTbRdMXcClW',
    'YEVPmnaTsPDi', 'fIDZziSjYomhpnsYZeNT', 'VGnmQicvPPuUIpOf', 'PyWteiXtEqUoNJqQr', 'PZmWeLMuYdpU', 'qdbJwtxGHGUpSmQpQGnG', 'aLXlmYaNXYc', 'aVEuMZsKEaUkzo', 'ZqOCUQpOUOGaZTe', 'CLffVVrESxecDjQQuAzR',
    'bhJslsrRIXBWZUaCOI', 'rwIWVTEyRMJoLQOprqk', 'dlCLBURIgVfMzuSdno', 'bJfycTLdmIAHvOLgHS', 'ZdECejfuaXW', 'lXvtPenkjnwEZTFMr', 'FPbpGVQcCpUVAVYIT', 'OgurXIDBYbQPqNzZyb', 'PfkHAlOYqGt', 'vebVRSWvYLFtM'
]
random_patterns = [f"{pat}.py" for pat in random_data]


import pathlib
import time

path = pathlib.Path(".")

t0 = time.perf_counter()

for pat in random_data:
    list(path.rglob(pat))

print(time.perf_counter() - t0)
```

</details>

On `main` on my machine, the benchmark runs in about 15.8s; with this PR, it runs in 16.1s.

You could get nearly all of that back if you did something like this, though (the diff is relative to your PR branch):

```diff
--- a/Lib/pathlib.py
+++ b/Lib/pathlib.py
@@ -72,18 +72,23 @@ def _is_case_sensitive(pathmod):
 # Globbing helpers
 #

+re = glob = None
+

 @functools.lru_cache(maxsize=256)
 def _compile_pattern(pat, sep, case_sensitive):
     """Compile given glob pattern to a re.Pattern object (observing case
     sensitivity)."""
-    from glob import translate
-    regex = translate(pat, recursive=True, include_hidden=True, seps=sep)
+    global re, glob
+    if re is None:
+        import re, glob
+
+    flags = re.NOFLAG if case_sensitive else re.IGNORECASE
+    regex = glob.translate(pat, recursive=True, include_hidden=True, seps=sep)
     # The string representation of an empty path is a single dot ('.'). Empty
     # paths shouldn't match wildcards, so we consume it with an atomic group.
     regex = r'(\.\Z)?+' + regex
-    from re import compile, NOFLAG, IGNORECASE
-    return compile(regex, flags=NOFLAG if case_sensitive else IGNORECASE).match
+    return re.compile(regex, flags=flags).match
```